pallets/staking/src/lib.rs

```diff
@@ -0,0 +1,926 @@
+use sp_runtime::{DispatchError, FixedPointNumber, FixedU128};
+use sp_std::num::NonZeroU128;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarks;
+
+pub mod integrations;
+pub mod traits;
+pub mod types;
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+/// Lock for staked amount by user
+pub const STAKING_LOCK_ID: LockIdentifier = *b"stk_stks";
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use crate::traits::{DemocracyReferendum, Freeze};
+	use crate::types::Voting;
+	use codec::HasCompact;
+	use frame_support::PalletId;
+	use frame_system::{ensure_signed, pallet_prelude::*};
+	use orml_traits::GetByKey;
+	use sp_runtime::traits::AtLeast32BitUnsigned;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Origin to initialize staking
+		type AuthorityOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Identifier for the class of asset.
+		type AssetId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ TypeInfo;
+
+		/// Multi currency mechanism.
+		type Currency: MultiLockableCurrency<Self::AccountId, CurrencyId = Self::AssetId, Balance = Balance>;
+
+		/// Staking period length in blocks.
+		#[pallet::constant]
+		type PeriodLength: Get<Self::BlockNumber>;
+
+		/// Pallet id.
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Native Asset ID
+		#[pallet::constant]
+		type NativeAssetId: Get<Self::AssetId>;
+
+		/// Min amount user must stake.
+		#[pallet::constant]
+		type MinStake: Get<Balance>;
+
+		/// Weight of the time points in total points calculations.
+		#[pallet::constant]
+		type TimePointsWeight: Get<Permill>;
+
+		/// Weight of the action points in total points calculations.
+		#[pallet::constant]
+		type ActionPointsWeight: Get<Permill>;
+
+		/// Number of time points users receive for each period.
+		#[pallet::constant]
+		type TimePointsPerPeriod: Get<u8>;
+
+		/// Number of periods user can't claim rewards for. User can exit but won't receive any rewards.
+		/// If he stay longer than `UnclaimablePeriods` he will receive rewards also for these periods.
+		#[pallet::constant]
+		type UnclaimablePeriods: Get<Period>;
+
+		/// Weight of the actual stake in slash points calculation. Bigger the value lower the calculated slash points.
+		#[pallet::constant]
+		type CurrentStakeWeight: Get<u8>;
+
+		/// Unit we are distributing action points for.
+		/// e.g if RewardedVoteUnit is 1HDX user will receive `x` action points per each voted 1 HDX.
+		#[pallet::constant]
+		type RewardedVoteUnit: Get<Balance>;
```

Comment:
Not necessarily,  amount of action points distributed per "unit" is also configurable. 
e.g it can be cofigrted to  distribute 4 action points per each voted 25HDX